### PR TITLE
Add Postgres backed MemoryStore with integration tests

### PR DESCRIPTION
**Pull request by aaryan-iiitu**
### Summary

This PR adds a Postgres backed implementation of `MemoryStore` which was mentioned in #363  to support persistent agent memory for production and long running workflows. 

### What’s included

- `PostgresMemoryStore` using JDBC + HikariCP
- Automatic schema initialization with pgvector support
- Integration tests using Testcontainers (Postgres + pgvector)
- Test scoped dependencies only (no runtime impact)

### Design notes
- This is an MVP focused on persistence (store / get / recall / delete).
- Semantic search and full `MemoryFilter` support will be added in the later PR's.
- Update operations are not yet concurrency safe and are documented as such.

### Testing
- Added integration tests that run against a real Postgres instance via Testcontainers.
- All existing tests pass.

### CODE REVIEW

**Comment by aaryan-iiitu** [comment]
@rorygraves  I have added unit tests for PostgresMemoryStore to cover core logic like SQL filters, metadata JSON handling, and embedding serialization.  Kindly take a look at it and please let me know if any modification is required .

**Comment by rorygraves** [comment]
@aaryan-iiitu Together with claude I have done a deep dive on this PR, you are off to a great start but there are a number of items to address (see above comment).

**Comment by aaryan-iiitu** [comment]
@rorygraves  Thanks for the detailed review. I've addressed the security issues (table name validation), switched to ujson for safe JSON handling, simplified the connection logic and implemented the missing filters. I also added unit tests for the edge cases you mentioned. Kindly inform me if any further changes is required .

File: modules/core/src/main/scala/org/llm4s/agent/memory/PostgresMemoryStore.scala [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+ * - This is an MVP implementation focused on persistence only.
+ * - Compound filters (And/Or/Not) and semantic search will be added later.
+ */
+final class PostgresMemoryStore private[memory] (
+  private val dataSource: HikariDataSource,
+  val tableName: String
+) extends MemoryStore
+    with AutoCloseable {
+
+  initializeSchema()
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/agent/memory/PostgresMemoryStore.scala:24]: side effect`initializeSchema()` runs in the constructor. Prefer making
initialization explicit in apply (factory) so construction stays referentially transparent and easier to test.

File: modules/core/src/main/scala/org/llm4s/agent/memory/PostgresMemoryStore.scala [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+          memory.embedding match {
+            case Some(vec) => stmt.setString(7, PostgresMemoryStore.embeddingToString(vec))
+            case None      => stmt.setNull(7, java.sql.Types.OTHER, "vector")
+          }
+
+          stmt.executeUpdate()
+        }
+      }
+      this
+    }.toEither.left.map(e => ProcessingError("postgres-memory-store", s"Failed to store memory: ${e.getMessage}"))
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/agent/memory/PostgresMemoryStore.scala:93]: broad Try {.. }.toEither.left.map(..) is OK at the JDBC boundary, but you are dropping the cause. Use
ProcessingError(.., cause = Some(e)) so callers do not lose stack traces (ProcessingError supports this).

File: modules/core/src/main/scala/org/llm4s/agent/memory/PostgresMemoryStore.scala [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+        throw new IllegalArgumentException(s"Invalid metadata key: '$key'")
+      }
+      (s"metadata->>'$key' = ?", Seq(value))
+
+    case unsupported =>
+      throw new UnsupportedOperationException(
+        s"PostgresMemoryStore does not yet support filter: ${unsupported.getClass.getSimpleName}"
+      )
+  }
+
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/agent/memory/PostgresMemoryStore.scala:317]: `filterToSql` uses throw for invalid metadata keys and unsupported filters. Even though its caught by outer Try, this is still
"exceptions for control flow". Prefer returning `Result[(String, Seq[...])]` (or a domain error ADT) and handle it explicitly in recall/deleteMatching/count.

File: modules/core/src/main/scala/org/llm4s/agent/memory/PostgresMemoryStore.scala [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+      content = rs.getString("content"),
+      memoryType = MemoryType.fromString(rs.getString("memory_type")),
+      metadata = PostgresMemoryStore.jsonToMetadata(rs.getString("metadata")),
+      timestamp = rs.getTimestamp("created_at").toInstant,
+      importance = Option(rs.getDouble("importance")).filterNot(_ => rs.wasNull()),
+      embedding = Option(embeddingStr).map(PostgresMemoryStore.stringToEmbedding)
+    )
+  }
+
+  private def setParameter(stmt: PreparedStatement, index: Int, value: Any): Unit = value match {
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/agent/memory/PostgresMemoryStore.scala:225]: `setParameter(.., value: Any)` Do not use `Any` this makes it type-unsafety. Suggest introducing a small internal
ADT (e.g., sealed trait SqlParam) so params are typed and exhaustively handled.

Comment (vim89) [reply modules/core/src/main/scala/org/llm4s/agent/memory/PostgresMemoryStore.scala:225]: SQLiteMemoryStore may already have the same Seq[Any] pattern; fixing it once in a shared helper would improve both.

File: modules/core/src/main/scala/org/llm4s/agent/memory/PostgresMemoryStore.scala [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+          params.zipWithIndex.foreach { case (param, idx) =>
+            setParameter(stmt, idx + 1, param)
+          }
+          stmt.setInt(params.size + 1, limit)
+
+          Using.resource(stmt.executeQuery()) { rs =>
+            val memories = ArrayBuffer.empty[Memory]
+            while (rs.next()) memories += rowToMemory(rs)
+            memories.toSeq
+          }
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/agent/memory/PostgresMemoryStore.scala:122]: imperative ArrayBuffer + while(rs.next()) works, but you already use a more idiomatic iterator pattern in SQLiteMemoryStore (Iterator.continually(rs).takeWhile(_.next())..) Consider using the same style for consistency.

File: modules/core/src/main/scala/org/llm4s/agent/memory/PostgresMemoryStore.scala [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+  }
+
+  private[memory] def metadataToJson(metadata: Map[String, String]): String =
+    if (metadata.isEmpty) "{}"
+    else ujson.write(metadata)
+
+  private[memory] def jsonToMetadata(json: String): Map[String, String] =
+    if (json == null || json == "{}" || json.isEmpty) Map.empty
+    else ujson.read(json).obj.map { case (k, v) => k -> v.str }.toMap
+
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/agent/memory/PostgresMemoryStore.scala:325]: `jsonToMetadata` assumes all JSON values are strings (v.str). If the DB contains non string JSON (even accidentally), this throws. Consider safer decoding (stringify non-strings or fail with a typed Left that includes context).

**Comment by aaryan-iiitu** [comment]
Thanks @vim89 @rorygraves  for the feedbacks. I have implemented all the review feedback locally and corrected the code. Would you prefer the review feedback addressed in one PR or split into a few smaller focused PRs ? I just wanted a clarification on this

**Comment by vim89** [comment]
> Thanks @vim89 @rorygraves for the feedbacks. I have implemented all the review feedback locally and corrected the code. Would you prefer the review feedback addressed in one PR or split into a few smaller focused PRs ? I just wanted a clarification on this

I'd recommend multiple small PRs

**Comment by rorygraves** [comment]
@aaryan-iiitu This is in much better shape.
Can you please address @vim89 FP style comments in the code and mark them as resolved once done i.e.:
These FP-style improvements haven't been addressed yet:

  1. Side effect in constructor (PostgresMemoryStore.scala:24)
  initializeSchema()  // Called in constructor - side effect
  Suggestion: Move to factory method for referential transparency.

  2. Error cause is dropped (PostgresMemoryStore.scala:93)
  .toEither.left.map(e => ProcessingError("...", s"Failed: ${e.getMessage}"))
  The stack trace is lost. Use ProcessingError(..., cause = Some(e)) if supported.

  3. Exceptions for control flow (PostgresMemoryStore.scala:304-308)
  case MemoryFilter.ByMetadata(key, value) =>
    if (!ValidIdentifierPattern.matches(key)) {
      throw new IllegalArgumentException(...)  // Throwing in helper method
    }
  Should return Result[(String, Seq[...])] instead.

  4. Type-unsafe Seq[Any] parameters (PostgresMemoryStore.scala:225)
  private def setParameter(stmt: PreparedStatement, index: Int, value: Any): Unit
  Suggest introducing a sealed trait SqlParam ADT for type safety.

  5. Imperative iteration style (PostgresMemoryStore.scala:122)
  val memories = ArrayBuffer.empty[Memory]
  while (rs.next()) memories += rowToMemory(rs)
  Consider using Iterator.continually(rs).takeWhile(_.next()) for consistency with SQLiteMemoryStore.

  6. Unsafe JSON parsing (PostgresMemoryStore.scala:325)
  ujson.read(json).obj.map { case (k, v) => k -> v.str }.toMap
  Assumes all values are strings - will throw if DB contains non-string JSON values. Should handle gracefully.


and then I think we are good to go.

**Comment by aaryan-iiitu** [comment]
@rorygraves  @vim89  I have addressed all the review comments and updated the unit tests accordingly.  Could you please take a look and let me know if anything is still missing or needs adjustment?
I grouped these changes into a single PR because they are closely related refactors of the same component , so it felt cleaner to change them together . I will surely split it further if you say so .

**Comment by rorygraves** [comment]
## Code Review: PR #371 - PostgresMemoryStore

Thanks for the updates @aaryan-iiitu! I've reviewed the latest code against our [review guidelines](https://github.com/llm4s/llm4s/blob/main/docs/reference/review-guidelines.md).

---

### ✅ Issues Successfully Addressed

| Guideline | Status | Notes |
|-----------|--------|-------|
| **#7 JSON Handling** | ✅ Fixed | Switched to `ujson` (lines 315-325) |
| **#8 Validation** | ✅ Fixed | Table name validation with regex (lines 259-264, 281-285) |
| **#8 Validation** | ✅ Fixed | Metadata key validation (lines 301-304) |
| **#4 Resource Management** | ✅ Good | Uses `Using.resource()` throughout |
| **#1 Error Handling** | ✅ Partial | Uses `Result[A]` for return types |

---

### 🔴 Outstanding FP-Style Issues

Based on @rorygraves and @vim89's previous feedback, these items still need attention:

#### 1. Constructor Side Effect (Guideline #5)
**Location:** `PostgresMemoryStore.scala:24`
```scala
initializeSchema()  // ❌ Side effect in constructor
```
**Fix:** Move schema initialization to the `apply` factory method for referential transparency.

---

#### 2. Error Causes Dropped (Guideline #1)
**Location:** Multiple (e.g., lines 93, 100, 148)
```scala
.toEither.left.map(e => ProcessingError("postgres-memory-store", s"Failed: ${e.getMessage}"))
```
Stack traces are valuable for debugging. Use:
```scala
.toEither.left.map(e => ProcessingError("postgres-memory-store", s"Failed: ${e.getMessage}", cause = Some(e)))
```

---

#### 3. Exceptions for Control Flow (Guideline #1)
**Location:** `PostgresMemoryStore.scala:301-309`
```scala
throw new IllegalArgumentException(...)  // ❌ in ByMetadata
throw new UnsupportedOperationException(...)  // ❌ in unsupported case
```
**Fix:** Change `filterToSql` to return `Result[(String, Seq[SqlParam])]` instead of throwing.

---

#### 4. Type-Unsafe `Any` (Guideline #3)
**Location:** `PostgresMemoryStore.scala:229`
```scala
private def setParameter(stmt: PreparedStatement, index: Int, value: Any): Unit
```
**Fix:** Use a sealed ADT:
```scala
sealed trait SqlParam
case class StringParam(value: String) extends SqlParam
case class IntParam(value: Int) extends SqlParam
case class TimestampParam(value: Timestamp) extends SqlParam
```

---

#### 5. Imperative Iteration Style (Guideline #15)
**Location:** `PostgresMemoryStore.scala:122-125`
```scala
val memories = ArrayBuffer.empty[Memory]
while (rs.next()) memories += rowToMemory(rs)
```
**More idiomatic:**
```scala
Iterator.continually(rs).takeWhile(_.next()).map(rowToMemory).toSeq
```

---

#### 6. Unsafe JSON Parsing (Guideline #9)
**Location:** `PostgresMemoryStore.scala:320-322`
```scala
ujson.read(json).obj.map { case (k, v) => k -> v.str }.toMap
```
This throws if any JSON value isn't a string. Handle gracefully:
```scala
ujson.read(json).obj.flatMap { case (k, v) => v.strOpt.map(k -> _) }.toMap
```

---

### Summary

The security issues from the initial review have been addressed. The remaining items are FP-style refinements. You mentioned you've implemented these locally - please push those changes and we should be good to merge!

**Comment by rorygraves** [comment]
@aaryan-iiitu  Sorry for the delay in reviews, my week has been crazy busy.
See above comment for details - this is looking really good.  Applying the new review guidelines - there are a few minor cleanups to address - then lets get this nice additional feature merged!.

**Comment by rorygraves** [comment]
Completed under #426
